src/xrpld/app/tx/detail/PermissionedDomainSet.cpp

```diff
@@ -0,0 +1,159 @@
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/detail/PermissionedDomainSet.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/protocol/STObject.h>
+#include <xrpl/protocol/TxFlags.h>
+#include <map>
+#include <optional>
+
+namespace ripple {
+
+NotTEC
+PermissionedDomainSet::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featurePermissionedDomains))
+        return temDISABLED;
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto const credentials = ctx.tx.getFieldArray(sfAcceptedCredentials);
+    if (credentials.empty() || credentials.size() > PD_ARRAY_MAX)
+        return temMALFORMED;
+    for (auto const& credential : credentials)
+    {
+        if (!credential.isFieldPresent(sfIssuer) ||
+            !credential.isFieldPresent(sfCredentialType))
+        {
+            return temMALFORMED;
+        }
+        auto sz = credential.getFieldVL(sfCredentialType).size();
+        if (!sz || sz > 64)
+            return temMALFORMED;
+    }
+
+    auto const domain = ctx.tx.at(~sfDomainID);
+    if (domain && *domain == beast::zero)
+        return temMALFORMED;
+
+    return preflight2(ctx);
+}
+
+TER
+PermissionedDomainSet::preclaim(PreclaimContext const& ctx)
+{
+    if (!ctx.view.read(keylet::account(ctx.tx.getAccountID(sfAccount))))
+        return tefINTERNAL;
+
+    auto const credentials = ctx.tx.getFieldArray(sfAcceptedCredentials);
+    for (auto const& credential : credentials)
+    {
+        if (!ctx.view.read(keylet::account(credential.getAccountID(sfIssuer))))
+            return temBAD_ISSUER;
+    }
+
+    if (!ctx.tx.isFieldPresent(sfDomainID))
+        return tesSUCCESS;
+    auto const domain = ctx.tx.getFieldH256(sfDomainID);
+    auto const sleDomain = ctx.view.read(keylet::permissionedDomain(domain));
+    if (!sleDomain)
+        return tecNO_ENTRY;
+    auto const owner = sleDomain->getAccountID(sfOwner);
+    auto account = ctx.tx.getAccountID(sfAccount);
+    if (owner != account)
+        return temINVALID_ACCOUNT_ID;
+
+    return tesSUCCESS;
+}
+
+/** Attempt to create the Permissioned Domain. */
+TER
+PermissionedDomainSet::doApply()
+{
+    auto const ownerSle = view().peek(keylet::account(account_));
+
+    // The purpose of this lambda is to modify the SLE for either creating a
+    // new or updating an existing object, to reduce code repetition.
+    // All checks have already been done. Just update credentials. Same logic
+    // for either new domain or updating existing.
+    // Silently remove duplicates.
```

Comment:
Should error for duplicates, not silently remove, since this change was made in the Credentials code.